File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/press_key.kt
package io.gopiper.piper.engine.statements

import android.util.Log
import android.view.KeyEvent
import io.gopiper.piper.engine.Piper
import io.gopiper.piper.engine.pressKey
import org.json.JSONObject

suspend fun Piper.executePressKey(statement: JSONObject, onError: (e: Exception) -> Unit) {
    try {
        Log.d("hello", "execute_press_key")
        when (statement.getString("pointer")) {
            "ENTER" -> pressKey(this.browser, KeyEvent.KEYCODE_ENTER, onError)
            "SHIFT" -> pressKey(this.browser, KeyEvent.KEYCODE_SHIFT_LEFT, onError)
            "CTRL" -> pressKey(this.browser, KeyEvent.KEYCODE_CTRL_LEFT, onError)
            "CAPS_LOCK" -> pressKey(this.browser, KeyEvent.KEYCODE_CAPS_LOCK, onError)
            "TAB" -> pressKey(this.browser, KeyEvent.KEYCODE_TAB, onError)
            "ESC" -> pressKey(this.browser, KeyEvent.KEYCODE_ESCAPE, onError)
        }
    } catch (e: Exception) {
        onError(e)
    }
}

File: piper/extensions/chrome/src/engine/statements/press_key.js
import { stringify, sleep } from "../util"


export default async function(statement, onError) {
    try {
        let tab = await this.browser.getCurrentTab()
        await chrome.debugger.attach({ tabId: tab.id }, "1.2")
        switch(statement.pointer) {
            case "ENTER":
                await chrome.debugger.sendCommand({ tabId: tab.id }, "Input.dispatchKeyEvent", {
                    type: "char",
                    text: String.fromCharCode(13),
                    modifiers: 0
                })
                break
            case "SHIFT":
                await chrome.debugger.sendCommand({ tabId: tab.id }, 'Input.dispatchKeyEvent', { 
                    type: "char",
                    text: String.fromCharCode(16), 
                    modifiers: 0
                })
                break
            case "CTRL":
                await chrome.debugger.sendCommand({ tabId: tab.id }, 'Input.dispatchKeyEvent', { 
                    type: "char",
                    text: String.fromCharCode(17), 
                    modifiers: 0
                })
                break
            case "CAPS_LOCK":
                await chrome.debugger.sendCommand({ tabId: tab.id }, 'Input.dispatchKeyEvent', { 
                    type: "char",
                    text: String.fromCharCode(20), 
                    modifiers: 0
                })
                break
            case "TAB":
                await chrome.debugger.sendCommand({ tabId: tab.id }, "Input.dispatchKeyEvent", {
                    type: "char",
                    text: String.fromCharCode(9),
                    modifiers: 0
                })
                break
            case "ESC":
                await chrome.debugger.sendCommand({ tabId: tab.id }, 'Input.dispatchKeyEvent', { 
                    type: "char",
                    text: String.fromCharCode(27), 
                    modifiers: 0
                })
                break
        }
        await chrome.debugger.detach({ tabId: tab.id })
        await sleep(1000)
    } catch (e) {
        console.log(e)
        onError(e)
    }
}
